R/shiny.R

```diff
@@ -53,7 +53,18 @@ renderDataTable = function(expr, env = parent.frame(), quoted = FALSE, ...) {
   )
   checkShinyVersion()
   if (!quoted) expr = substitute(expr)
-  htmlwidgets::shinyRenderWidget(expr, dataTableOutput, env, quoted = TRUE)
+
+  exprFunc <- shiny::exprToFunction(expr, env, quoted = TRUE)
+  widgetFunc <- function() {
+    x <- exprFunc()
+    if (is.data.frame(x)) {
+      datatable(x)
+    } else {
+      x
+    }
```

Comment:
Oh, it's trickier than that (but still possible). I'll put together a separate PR to demonstrate.